src/bases/D3MPoolBase.sol

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Copyright (C) 2021-2022 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.12;
+
+interface TokenLike {
+    function approve(address, uint256) external returns (bool);
+    function transfer(address, uint256) external returns (bool);
+    function transferFrom(address, address, uint256) external returns (bool);
+    function balanceOf(address) external view returns (uint256);
+}
+
+interface CanLike {
+    function hope(address) external;
+    function nope(address) external;
+}
+
+interface d3mHubLike {
+    function vat() external view returns (address);
+}
+
+abstract contract D3MPoolBase {
+    // --- Auth ---
+    mapping (address => uint256) public wards;
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+    modifier auth {
+        require(wards[msg.sender] == 1, "D3MPoolBase/not-authorized");
+        _;
+    }
+
+    TokenLike   public immutable asset; // Dai
+    address     public immutable hub;
+
+    address     public           share;
```

Comment:
I think we can drop the concept of the `share` and replace it with pool-specific tokens such as `adai` in Aave and `cdai` in Compound. Overloading these terms too much can be confusing. It's also entirely possible that an adapter does not have an associated claim token attached to it.

Reply:
that sounds good to.  I am a little confused though if the hub calls this on an adapter without an underlying share, what do we return? just the assets?

Reply:
Can return output same as input I guess.

Reply:
So the pool adapter is there to abstract away the design of the pool. Most pools issue an ERC20 claim token when you deposit, but they don't necessarily have to, and even if they do the semantics of the token is not the same for example aDAI rebases the `balanceOf()` whereas cDAI doesn't. The `convertToShares()` function can still work even if there is no ERC20 token attached you just need to convert between some fixed number and the rebase interest-earning number.

Reply:
:+1:  https://github.com/makerdao/dss-direct-deposit/pull/31/commits/2d808ff9d977fe8645b6dc7c6a3a0ab144717f5e

Reply:
In the case that a pool doesn't issue an ERC20 token, what are we providing to users in the case of ES?